### ISSUE TITLE
请问merge细节在哪里实现

### ISSUE THREAD

**Issue by YZW-explorer** [HuangOwen/RoLoRA#3]

**Comment by HuangOwen** [comment]
Please refer to this function https://github.com/HuangOwen/RoLoRA/blob/0118cb2decbcdeb5264f2374a53ad592e521c0a4/src/llamafactory/train/tuner.py#L50

**Comment by HuangOwen** [comment]
For merging LoRA with base model, we use https://github.com/HuangOwen/RoLoRA/blob/main/merge_peft_adapter.py

**Comment by YZW-explorer** [comment]
Hello author, first of all, I must admit that this is a very beautiful and valuable work, but I encountered a problem when reproducing your experiment in the paper.
My experiment shows that RoLoRA is not as effective as LoRA, and I am not sure where the problem lies.
Here is my experimental table
![image](https://github.com/user-attachments/assets/d837fb38-dd4c-4624-b241-e27b34a59053)
The following is the experimental curve
**RoLoRA/saves/llama2-7b-rolora-qv-r16/training_eval_loss.png**
![image](https://github.com/user-attachments/assets/70958c12-38a2-4ab9-a8ea-5208ebca4e4f)
**RoLoRA/saves/llama2-7b-rolora-qv-r16/training_loss.png**
![image](https://github.com/user-attachments/assets/32a46d95-418a-4562-99a4-61d6bb9130e7)
**RoLoRA/saves/llama2-7b-lora-qv-r16/training_eval_loss.png**
![image](https://github.com/user-attachments/assets/5909bd71-3a25-47a8-86b2-0e097f98755d)
**RoLoRA/saves/llama2-7b-lora-qv-r16/training_loss.png**
![image](https://github.com/user-attachments/assets/94e285bb-9bd5-4c91-b861-c1bb8a6ee15f)
I implemented it step by step according to the README without modifying any other variables.
**RoLoRA/rolora.sh**
CUDA_VISIBLE_DEVICES=5 python src/train.py \
--stage sft \
--do_train True \
--model_name_or_path meta-llama/Llama-2-7b-hf \
--rotate_down_proj \
--rotate_mode 'hadamard' \
--finetuning_type lora \
--template default \
--dataset alpaca_gpt4_en \
--cutoff_len 1024 \
--learning_rate 0.0001 \
--num_train_epochs 3.0 \
--max_samples 100000 \
--per_device_train_batch_size 8 \
--lr_scheduler_type cosine \
--max_grad_norm 1.0 \
--logging_steps 10 \
--save_steps 5000 \
--warmup_ratio 0.01 \
--val_size 0.1  \
--per_device_eval_batch_size 16 \
--evaluation_strategy steps \
--eval_steps 5000 \
--optim adamw_torch \
--report_to wandb \
--output_dir saves/llama2-7b-rolora-qv-r16 \
--fp16 True \
--lora_rank 16 \
--lora_alpha 16 \
--lora_dropout 0 \
--lora_target q_proj.module,v_proj.module \
--plot_loss True \
--load_best_model_at_end
**RoLoRA/merge_rolora.sh**
CUDA_VISIBLE_DEVICES=0 python src/export.py \
--model_name_or_path meta-llama/Llama-2-7b-hf \
--rotate_down_proj \
--rotate_mode 'hadamard' \
--adapter_name_or_path saves/llama2-7b-rolora-qv-r16/ \
--template default \
--finetuning_type lora \
--export_dir models/llama2-7b-rolora-qv-r16 \
--export_size 2 \
--export_device cpu \
--export_legacy_format false 
Looking forward to your reply, thank you!

**Comment by HuangOwen** [comment]
Thanks for your interest in our work! Your running scripts are correct and the training/eval loss also seems reasonable. I would recommend you check the dependencies (making sure all required packages are aligned with https://github.com/HuangOwen/RoLoRA/blob/main/requirements.txt ). In addition, please make sure to use our own evaluation scripts `sh eval_rolora.sh`, which is different from base lm-eval-harness.

**Comment by YZW-explorer** [comment]
You're right, it's indeed a problem with my environment. My experimental results are consistent with those in your paper, which is truly an amazing result, great👍!
May I ask how I can draw the visual distribution map in your paper to obtain a clearer and more detailed observation.

**Comment by HuangOwen** [comment]
Glad to hear that you can reproduce the results! For the distribution visualization, please refer to this script: https://gist.github.com/HuangOwen/4abdf5455dd14cb8901a701c921d25b2 .

**Comment by YZW-explorer** [comment]
Thank you very much!👍
